### PR TITLE
Optimize SDK initialization when requests executed before any activity starts

### PR DESCRIPTION
**Pull request by tonidero**
### Description
We discovered that the SDK might take a while to respond to offerings/customer info operations if they are started before any `Activity` has had the chance to start. This is because the SDK considers the app to be in the background, and we have some optimizations in place where we add some jittering to backend requests performed in the background

The approach in this PR is to make it so:
- If a request without a jitter delay is started after a request with jitter (but before it's completed), we move the callbacks to use the response from the request without jittering.
- If a request with jittering is started after a request without jitter (but before it's completed), we will not add the request with jitter but just reuse the existing request without jittering.

Note that with this approach, we might be duplicating requests in these situations... But I think it should be ok, lmk if you think otherwise

### CODE REVIEW

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt [hunk 1]
```diff
@@ -94,6 +95,7 @@ internal class OfferingsManager(
             },
             onSuccess = { offerings ->
                 offeringsCache.cacheOfferings(offerings, offeringsJSON)
+                offeringsCache.setOfferingsCacheTimestampToNow()
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:98]: This is another optimization I had to do... If we left it where it was before, we wouldn't use the new mechanisms to optimize the requests trying to use the request that is supposed to finish earlier.

File: purchases/src/main/kotlin/com/revenuecat/purchases/subscriberattributes/SubscriberAttributesPoster.kt [hunk 2]
```diff
@@ -25,6 +26,7 @@ internal class SubscriberAttributesPoster(
             Endpoint.PostAttributes(appUserID),
             mapOf("attributes" to attributes),
             postFieldsToSign = null,
+            delay = Delay.DEFAULT,
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/subscriberattributes/SubscriberAttributesPoster.kt:29]: This is another change that I thought was fine but lmk if you have any concerns... Basically, this will add jittering to the subscriber attributes requests, which didn't exist before. I don't think it's critical so it made sense to me to delay it in favor of other requests.

**Review by tonidero — COMMENTED** [review]
I haven't started on tests, since I wanted to gather some feedback first

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt [hunk 3]
```diff
@@ -510,6 +513,23 @@ internal class Backend(
...
 
+    private fun <S, E> MutableMap<BackgroundAwareCallbackCacheKey, MutableList<Pair<S, E>>>.addBackgroundAwareCallback(
+        call: Dispatcher.AsyncCall,
+        dispatcher: Dispatcher,
+        cacheKey: BackgroundAwareCallbackCacheKey,
+        functions: Pair<S, E>,
+        delay: Delay = Delay.NONE,
+    ) {
+        val foregroundCacheKey = Pair(cacheKey.first, false)
+        val cacheKeyToUse = if (cacheKey.second && callbacks.containsKey(foregroundCacheKey)) {
```
[the remark this answers is not in the document]
Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:524]: Also maybe extract `callbacks.containsKey(foregroundCacheKey)` into a boolean with a descriptive name, something like `foregroundCallAlreadyInPlace`

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt [hunk 4]
```diff
@@ -510,6 +513,23 @@ internal class Backend(
...
+            foregroundCacheKey
+        } else {
+            cacheKey
+        }
+        addCallback(call, dispatcher, cacheKeyToUse, functions, delay)
+        moveBackgroundedCallbacksToForegroundedCallbacksIfNeeded(this, cacheKey.first, cacheKey.second)
```
Comment (vegaro) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:530]: It's only necessary to do this if this new callback being added is unjittered right?

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:524]: > can you create variables for cacheKey.first and cacheKey.second? I think it's a bit hard to follow what's what

Hmm yeah, I might create a data class to have these 2 so it's clearer 👍

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt (hunk 4 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:530]: Yes, that's correct. We do that check inside the method, but we might be able to move it here.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt [hunk 5]
```diff
@@ -61,11 +61,12 @@ internal class OfferingsManager(
         onError: ((PurchasesError) -> Unit)? = null,
         onSuccess: ((Offerings) -> Unit)? = null,
     ) {
-        offeringsCache.setOfferingsCacheTimestampToNow()
```
Comment (vegaro) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:64]: I don't recall exactly why this was here before 🤔 But I remember there was a reason

Brain dump:

Before your changes, if there was another request to fetch offerings, the request considers the cache to be recent and it would return that outdated offerings cache instead, but that's incorrect right? I think that's what you optimized.

Also, if there was an issue getting the offerings, we would clear that newly set timestamp, and the following request would retry, with your changes, the behavior is the same

Anyways, I think the optimization makes sense, I am just trying to remember why this was done that way lol

Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:64]: >  if there was another request to fetch offerings, the request considers the cache to be recent and it would return that outdated offerings cache instead, but that's incorrect right?

Mostly yes. However, the scenario I was trying to fix was the following:
- User calls `getOfferings` BEFORE` activity `onStart`. We mark timestamp of the cache to now before the requests starts
- Activity `onStart` happens, however, the offerings cache already has a recent timestamp so we don't even try to get it (since it happens in the background, we don't expect a result, so we don't even check if there is a cached value or not).
- We remain only with the request that had the jittering, rendering the main changes in this PR useless.

> Also, if there was an issue getting the offerings, we would clear that newly set timestamp, and the following request would retry, with your changes, the behavior is the same

That's correct, and that remains that way for now, though it probably wouldn't be needed now. I've just made it so marking the offering cache timestamp happens only after a success.

The main consequence of this change is that multiple offerings request may happen before we receive a successful response, however, we rely on the request callback cache in the `Backend` to avoid duplicate requests.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt]: I created a data class to make things clearer here: https://github.com/RevenueCat/purchases-android/pull/1204/commits/6aa761f05530ab99af09918c162165809d41a83e, also extracted those conditions to variables to make the code more legible.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt (hunk 4 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt]: Removed that function and moved it to the main one, so it's clearer in https://github.com/RevenueCat/purchases-android/pull/1204/commits/6aa761f05530ab99af09918c162165809d41a83e

**Review by vegaro — APPROVED** [review]

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt [hunk 6]
```diff
@@ -510,6 +512,39 @@ internal class Backend(
...
+        addCallback(call, dispatcher, cacheKeyToUse, functions, delay)
+        // In case we have a request with a jittered delay queued, and we perform the same request without
+        // jittered delay, we want to call the callback using the unjittered request
+        val backgroundedCacheKey = cacheKey.copy(appInBackground = true)
+        val backgroundCallAlreadyInPlace = containsKey(foregroundCacheKey)
+        if (!cacheKey.appInBackground && backgroundCallAlreadyInPlace) {
+            warnLog(NetworkStrings.SAME_CALL_SCHEDULED_WITH_JITTER.format(foregroundCacheKey))
+            remove(backgroundedCacheKey)?.takeIf { it.isNotEmpty() }?.let { backgroundedCallbacks ->
+                if (containsKey(cacheKey)) {
+                    this[cacheKey]?.addAll(backgroundedCallbacks)
```
Comment (aboedo) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:540]: 👏

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt (hunk 5 above)
Comment (aboedo) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:64]: We had a recent conversation about this on iOS as well. Using the "set timestamp to now" workaround is effective but definitely not intuitive and confusing for maintainers. 

I think the ultimate solution to it is to have 2 values - "last updated" and "is currently updating", and use the combination of those 2 as a more expressive way of determining state. We're currently encoding both of those into a single boolean value. 

The cache key came later, though, so maybe it removes the need for that workaround entirely? Does it de-duplicate requests with the same key? If so it effectively removes the need for the "set timestamp to now" workaround and even having an "is currently updating" value as well, right?

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt (hunk 1 above)
Comment (aboedo) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:98]: I wonder if these two lines should just be a part of cacheOfferings in offeringsCache and guarantee atomicity there

Comment (aboedo) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:98]: Feels like it'd be conceptually cleaner and slightly easier to make it synchronized?

File: purchases/src/test/java/com/revenuecat/purchases/common/BackendTest.kt [hunk 7]
```diff
@@ -335,6 +336,60 @@ class BackendTest {
         }
     }
 
+    @Test
+    fun `given getCustomerInfo call on foreground, then one in background, only one request without delay is triggered`() {
```
Comment (aboedo) [inline purchases/src/test/java/com/revenuecat/purchases/common/BackendTest.kt:340]: we have to do this because it'd be tricky to cancel the existing request for the background, right?

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt (hunk 5 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:64]: > The cache key came later, though, so maybe it removes the need for that workaround entirely? Does it de-duplicate requests with the same key?

Yes, the callback cache mechanism deduplicates requests while the request "is currently updating". However, we still need to store the last time the cache was updated to know whether it's stale, so we still need to store the timestamp on a success.

> We had a recent conversation about this on iOS as well. 

Yes, iOS did a similar change not too long ago and it relies on the request cache as well. I think that makes the most sense to handle deduplicating requests while it's in progress.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt (hunk 1 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:98]: Certainly, these 2 could be joined (this is the only place they are both called) I will do so in a different PR.
